src/utils/use-product-handler.js

```diff
@@ -0,0 +1,239 @@
+import { useState, useMemo, useCallback } from 'react';
+import { useSelector } from 'react-redux';
+import { config } from '../configs';
+
+const {
+  selectedLanguages,
+  languages,
+  product: { optionsValues }
+} = config;
+
+const useProductHandler = () => {
+  const { filterData, productOptions, modelsForSelectedCategory } = useSelector(
+    ({ Products }) => ({
+      filterData: Products.filterData,
+      productOptions: Products.productOptions,
+      modelsForSelectedCategory:
+        Products.productSpecies.modelsForSelectedCategory
+    })
+  );
+
+  const [preferedLanguages, setPreferedLanguages] = useState(selectedLanguages);
+  const [selectedOptions, setOptions] = useState(optionsValues);
+  const [primaryImage, setPrimaryImage] = useState('');
+  const [additionalImages, setAdditionalImages] = useState([]);
+
+  const { bottomMaterials: materials, sizes } = productOptions;
+
+  const checkedLanguages = useMemo(
+    () => Object.values(preferedLanguages).filter(({ checked }) => checked),
+    [preferedLanguages]
+  );
+
+  const categoriesNames = useMemo(
+    () => [
+      ...new Set(filterData.map(({ category }) => category.name[0].value))
+    ],
+    [filterData]
+  );
+
+  const categories = useMemo(
+    () =>
+      categoriesNames.map(
+        (category) =>
+          filterData.find(
+            ({ category: { name } }) => category === name[0].value
+          ).category
+      ),
+    [filterData, categoriesNames]
+  );
+
+  const colorsNames = useMemo(
+    () => [
+      ...new Set(filterData.map(({ colors }) => colors[0].simpleName[0].value))
+    ],
+    [filterData]
+  );
+
+  const colors = useMemo(
+    () =>
+      colorsNames.map(
+        (color) =>
+          filterData.find(
+            ({ colors }) => colors[0].simpleName[0].value === color
+          ).colors
+      ),
+    [filterData, colorsNames]
+  );
+
+  const patternsNames = useMemo(
+    () => [...new Set(filterData.map(({ pattern }) => pattern[0].value))],
+    [filterData]
+  );
+
+  const patterns = useMemo(
+    () =>
+      patternsNames.map(
+        (item) =>
+          filterData.find(({ pattern }) => pattern[0].value === item).pattern
+      ),
+    [filterData, patternsNames]
+  );
+
+  const modelNames = useMemo(
+    () => [...new Set(filterData.map(({ model }) => model[0].value))],
+    [filterData]
+  );
+
+  const models = useMemo(
+    () =>
+      modelNames.map(
+        (item) => filterData.find(({ model }) => model[0].value === item).model
+      ),
+    [filterData, modelNames]
+  );
+
+  const additions = useMemo(
+    () =>
+      filterData.length
+        ? filterData
+              .find(({ options }) => options.find(({ additions }) => additions.length))
+              .options.find(({ additions }) => additions.length)
+          .additions
+        : null,
+    [filterData]
+  );
+
+  const sizeOptions = useMemo(
+    () =>
+      selectedOptions.sizes.length
```

Comment:
There is no need to check for the length, in the result of `.map` we are going to get the array (empty or with items) anyway

Reply:
removed